src/target.rs

```diff
@@ -0,0 +1,190 @@
+use arrayvec::ArrayVec;
+use std::io::{BufRead, Cursor};
+use std::iter::IntoIterator;
+use std::ops::Deref;
+use std::process::Output;
+use tokio::process::Command;
+
+/// Compiled target triple, used as default for binary fetching
+pub const TARGET: &str = env!("TARGET");
+
+#[derive(Debug, Clone)]
+pub struct Targets(ArrayVec<Box<str>, 2>);
+
+impl Targets {
+    fn from_array<const LEN: usize>(arr: [Box<str>; LEN]) -> Self {
+        let mut v = ArrayVec::new();
+
+        for elem in arr {
+            v.push(elem);
+        }
+
+        Self(v)
+    }
+
+    fn push(&mut self, s: Box<str>) {
+        self.0.push(s)
+    }
+}
+
+impl Deref for Targets {
+    type Target = [Box<str>];
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl IntoIterator for Targets {
+    type Item = Box<str>;
+    type IntoIter = arrayvec::IntoIter<Box<str>, 2>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+/// Detect the targets supported at runtime,
+/// which might be different from `TARGET` which is detected
+/// at compile-time.
+///
+/// Return targets supported in the order of preference.
+/// If target_os is linux and it support gnu, then it is preferred
+/// to musl.
+///
+/// If target_os is mac and it is aarch64, then aarch64 is preferred
+/// to x86_64.
+///
+/// Check [this issue](https://github.com/ryankurte/cargo-binstall/issues/155)
+/// for more information.
+pub async fn detect_targets() -> Targets {
+    if let Some(target) = get_targets_from_rustc().await {
+        let mut v = Targets::from_array([target]);
+
+        #[cfg(target_os = "linux")]
+        if v[0].contains("gnu") {
+            v.push(v[0].replace("gnu", "musl").into_boxed_str());
+        }
+
+        #[cfg(target_os = "macos")]
+        if &*v[0] == macos::AARCH64 {
+            v.push(macos::X86.into());
+        }
+
+        v
+    } else {
+        #[cfg(target_os = "linux")]
+        {
+            linux::detect_targets_linux().await
+        }
+        #[cfg(target_os = "macos")]
+        {
+            macos::detect_targets_macos()
+        }
+        #[cfg(not(any(target_os = "linux", target_os = "macos")))]
+        {
+            Targets::from_array([TARGET.into()])
+        }
+    }
+}
+
+// Figure out what the host target is, from rustc or from this program's own build target
+async fn get_targets_from_rustc() -> Option<Box<str>> {
+    match Command::new("rustc").arg("-vV").output().await {
```

Comment:
I think that match can be rewritten as a combinator, like `.await.ok()?` or something

Reply:
Done!